File: Leetcode/3446 - Sort Matrix by Diagonals/solution.ts
function sortMatrix(grid: number[][]): number[][] {
    const n: number = grid.length;

    for (let x: number = 1; x < n; x++) {
        let elements: number[] = [];
        let i: number = 0, j: number = x;
        while (i < n && j < n) {
            elements.push(grid[i][j]);
            i++;
            j++;
        }
        elements.sort((a, b) => a - b);
        i = 0; j = x;
        let k: number = 0;
        while (i < n && j < n) {
            grid[i][j] = elements[k++];
            i++;
            j++;
        }
    }

    for (let x: number = 0; x < n; x++) {
        let elements: number[] = [];
        let i: number = x, j: number = 0;
        while (i < n && j < n) {
            elements.push(grid[i][j]);
            i++;
            j++;
        }
        elements.sort((a, b) => b - a);
        i = x; j = 0;
        let k: number = 0;
        while (i < n && j < n) {
            grid[i][j] = elements[k++];
            i++;
            j++;
        }
    }

    return grid;
};

File: Leetcode/3446 - Sort Matrix by Diagonals/solution.cpp
vector<vector<int>> sortMatrix(vector<vector<int>>& grid) {
    int n = grid.size();

    for (int x = 1; x < n; x++) {
        vector<int> elements;
        int i = 0, j = x;
        while (i < n && j < n) {
            elements.push_back(grid[i][j]);
            i++; j++;
        }
        sort(elements.begin(), elements.end());
        i = 0; j = x;
        int k = 0;
        while (i < n && j < n) {
            grid[i][j] = elements[k++];
            i++; j++;
        }
    }

    for (int x = 0; x < n; x++) {
        vector<int> elements;
        int i = x, j = 0;
        while (i < n && j < n) {
            elements.push_back(grid[i][j]);
            i++; j++;
        }
        sort(elements.begin(), elements.end(), greater<int>());
        i = x; j = 0;
        int k = 0;
        while (i < n && j < n) {
            grid[i][j] = elements[k++];
            i++; j++;
        }
    }

    return grid;
}